File: frontend/src/types/Paciente.ts
export interface Paciente {
  id?: number;
  nomeCompleto: string;
  nomeSocial?: string;
  nomeMae?: string;
  cpf: string;
  justificativaAusenciaCpf?: string;
  cns?: string;
  cartaoSus?: string; // Para compatibilidade com código existente
  sexo?: string;
  dataNascimento: string;
  acamado?: boolean;
  domiciliado?: boolean;
  condSaudeMental?: boolean;
  usaPlantas?: boolean;
  outrasCondicoes?: string;

  // Campos legados para compatibilidade
  nome?: string; // Para compatibilidade com código existente
  telefone?: string;
  email?: string;
  endereco?: string;
  cidade?: string;
  estado?: string;
  genero?: string;
  estadoCivil?: string;
  profissao?: string;
  contatoEmergencia?: string;
  telefoneEmergencia?: string;
  observacoesMedicas?: string;
  prioridade?: string;

  // Endereço
  municipio?: string;
  cep?: string;
  logradouro?: string;
  numero?: string;
  bairro?: string;
  complemento?: string;

  // Contato
  telefoneCelular?: string;
  telefoneContato?: string;

  // Documentos
  tipoSanguineo?: string;
  rg?: string;
  orgaoEmissor?: string;
  certidaoNascimento?: string;
  carteiraTrabalho?: string;
  tituloEleitor?: string;

  // Outros dados
  prontuarioFamiliar?: string;
  corRaca?: string;
  etnia?: string;
  escolaridade?: string;
  situacaoFamiliar?: string;
  cbo?: string;

  // Status
  ativo?: boolean;

  // Campos de auditoria
  dataCriacao?: string;
  dataAtualizacao?: string;
  criadoPor?: string;
  atualizadoPor?: string;
}

// Tipo para criação de paciente (campos obrigatórios)
export interface CriarPacienteRequest {
  // Campos obrigatórios
  nomeCompleto: string;
  dataNascimento: string;

  // Campos opcionais
  nomeSocial?: string;
  nomeMae?: string;
  cpf?: string;
  justificativaAusenciaCpf?: string;
  cns?: string;
  sexo?: string;
  acamado?: boolean;
  domiciliado?: boolean;
  condSaudeMental?: boolean;
  usaPlantas?: boolean;
  outrasCondicoes?: string;
  municipio?: string;
  cep?: string;
  logradouro?: string;
  numero?: string;
  bairro?: string;
  complemento?: string;
  telefoneCelular?: string;
  telefoneContato?: string;
  tipoSanguineo?: string;
  rg?: string;
  orgaoEmissor?: string;
  certidaoNascimento?: string;
  carteiraTrabalho?: string;
  tituloEleitor?: string;
  prontuarioFamiliar?: string;
  corRaca?: string;
  etnia?: string;
  escolaridade?: string;
  situacaoFamiliar?: string;
  cbo?: string;

  // Campos legados para compatibilidade
  nome?: string;
  cartaoSus?: string;
  telefone?: string;
  email?: string;
  endereco?: string;
  cidade?: string;
  estado?: string;
  genero?: string;
  estadoCivil?: string;
  profissao?: string;
  contatoEmergencia?: string;
  telefoneEmergencia?: string;
  observacoesMedicas?: string;
}

// Tipo para atualização de paciente (todos os campos opcionais exceto ID)
export interface AtualizarPacienteRequest {
  id: number;

  // Campos opcionais
  nomeCompleto?: string;
  nomeSocial?: string;
  nomeMae?: string;
  cpf?: string;
  justificativaAusenciaCpf?: string;
  cns?: string;
  sexo?: string;
  dataNascimento?: string;
  acamado?: boolean;
  domiciliado?: boolean;
  condSaudeMental?: boolean;
  usaPlantas?: boolean;
  outrasCondicoes?: string;
  municipio?: string;
  cep?: string;
  logradouro?: string;
  numero?: string;
  bairro?: string;
  complemento?: string;
  telefoneCelular?: string;
  telefoneContato?: string;
  tipoSanguineo?: string;
  rg?: string;
  orgaoEmissor?: string;
  certidaoNascimento?: string;
  carteiraTrabalho?: string;
  tituloEleitor?: string;
  prontuarioFamiliar?: string;
  corRaca?: string;
  etnia?: string;
  escolaridade?: string;
  situacaoFamiliar?: string;
  cbo?: string;
  ativo?: boolean;

  // Campos legados para compatibilidade
  nome?: string;
  cartaoSus?: string;
  telefone?: string;
  email?: string;
  endereco?: string;
  cidade?: string;
  estado?: string;
  genero?: string;
  estadoCivil?: string;
  profissao?: string;
  contatoEmergencia?: string;
  telefoneEmergencia?: string;
  observacoesMedicas?: string;
}

// Tipo para resposta de busca
export interface PacienteBuscaResponse {
  pacientes: Paciente[];
  total: number;
  pagina: number;
  totalPaginas: number;
}

File: backend/src/main/java/com/sistemadesaude/backend/paciente/entity/Paciente.java
package com.sistemadesaude.backend.paciente.entity;

import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.time.LocalDate;
import java.time.LocalDateTime;

/**
 * Entidade que representa um Paciente no sistema
 */
@Entity
@Table(name = "pacientes")
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@ToString(exclude = {"endereco"}) // Exclui relacionamentos lazy
@EqualsAndHashCode(onlyExplicitlyIncluded = true)

public class Paciente {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "nome_completo", nullable = false, length = 200)
    private String nomeCompleto;

    @Column(name = "nome_social", length = 200)
    private String nomeSocial;

    @Column(name = "nome_mae", length = 200)
    private String nomeMae;

    @Column(name = "cpf", length = 11, unique = true)
    private String cpf;

    @Column(name = "justificativa_ausencia_cpf", length = 500)
    private String justificativaAusenciaCpf;

    @Column(name = "cns", length = 15)
    private String cns;

    @Column(name = "sexo", length = 1)
    private String sexo;

    @Column(name = "data_nascimento")
    private LocalDate dataNascimento;

    @Column(name = "acamado")
    private Boolean acamado;

    @Column(name = "domiciliado")
    private Boolean domiciliado;

    @Column(name = "cond_saude_mental")
    private Boolean condSaudeMental;

    @Column(name = "usa_plantas")
    private Boolean usaPlantas;

    @Column(name = "outras_condicoes", length = 1000)
    private String outrasCondicoes;

    @Column(name = "municipio", length = 100)
    private String municipio;

    @Column(name = "cep", length = 8)
    private String cep;

    @Column(name = "logradouro", length = 200)
    private String logradouro;

    @Column(name = "numero", length = 10)
    private String numero;

    @Column(name = "bairro", length = 100)
    private String bairro;

    @Column(name = "complemento", length = 100)
    private String complemento;

    @Column(name = "telefone_celular", length = 15)
    private String telefoneCelular;

    @Column(name = "telefone_contato", length = 15)
    private String telefoneContato;

    @Column(name = "tipo_sanguineo", length = 3)
    private String tipoSanguineo;

    @Column(name = "rg", length = 20)
    private String rg;

    @Column(name = "orgao_emissor", length = 10)
    private String orgaoEmissor;

    @Column(name = "certidao_nascimento", length = 50)
    private String certidaoNascimento;

    @Column(name = "carteira_trabalho", length = 20)
    private String carteiraTrabalho;

    @Column(name = "titulo_eleitor", length = 15)
    private String tituloEleitor;

    @Column(name = "prontuario_familiar", length = 20)
    private String prontuarioFamiliar;

    @Column(name = "cor_raca", length = 50)
    private String corRaca;

    @Column(name = "etnia", length = 50)
    private String etnia;

    @Column(name = "escolaridade", length = 50)
    private String escolaridade;

    @Column(name = "situacao_familiar", length = 100)
    private String situacaoFamiliar;

    @UpdateTimestamp
    @Column(name = "data_atualizacao")
    private LocalDateTime dataAtualizacao;

    /**
     * Método auxiliar para obter o endereço como um objeto de valor.
     * Este método normaliza os campos de endereço em um objeto Endereco.
     */
    @Transient
    public Endereco getEndereco() {
        return Endereco.builder()
                .municipio(this.municipio)
                .cep(this.cep)
                .logradouro(this.logradouro)
                .numero(this.numero)
                .bairro(this.bairro)
                .complemento(this.complemento)
                .build();
    }

    /**
     * Método auxiliar para definir o endereço a partir de um objeto de valor.
     * Este método desnormaliza o objeto Endereco nos campos da entidade.
     */
    @Transient
    public void setEndereco(Endereco endereco) {
        if (endereco != null) {
            this.municipio = endereco.getMunicipio();
            this.cep = endereco.getCep();
            this.logradouro = endereco.getLogradouro();
            this.numero = endereco.getNumero();
            this.bairro = endereco.getBairro();
            this.complemento = endereco.getComplemento();
        }
    }

    /**
     * Retorna o nome de exibição (nome social se disponível, senão nome completo)
     */
    @Transient
    public String getNomeExibicao() {
        return nomeSocial != null && !nomeSocial.trim().isEmpty() ? nomeSocial : nomeCompleto;
    }

    /**
     * Verifica se o paciente possui CPF válido
     */
    @Transient
    public boolean temCpfValido() {
        return cpf != null && !cpf.trim().isEmpty() && cpf.length() == 11;
    }

    /**
     * Calcula a idade aproximada baseada na data de nascimento
     */
    @Transient
    public Integer getIdade() {
        if (dataNascimento == null) return null;
        return LocalDate.now().getYear() - dataNascimento.getYear();
    }
}